File: Subarray with Given Sum 32/main.java
import java.util.*;

public class Solution {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        long k = sc.nextLong();

        long[] arr = new long[n];
        for (int i = 0; i < n; i++) {
            arr[i] = sc.nextLong();
        }

        HashSet<Long> prefix = new HashSet<>();
        long sum = 0;

        for (int i = 0; i < n; i++) {
            sum += arr[i];

            if (sum == k || prefix.contains(sum - k)) {
                System.out.println("YES");
                return;
            }

            prefix.add(sum);
        }

        System.out.println("NO");
    }
}

File: Subarray with Given Sum 32/main.cpp


#include <bits/stdc++.h>
using namespace std;

int main() {
    int n;
    long long k;
    cin >> n >> k;

    vector<long long> arr(n);
    for (int i = 0; i < n; i++) cin >> arr[i];

    unordered_set<long long> prefix;
    long long sum = 0;

    for (int i = 0; i < n; i++) {
        sum += arr[i];

        if (sum == k || prefix.count(sum - k)) {
            cout << "YES";
            return 0;
        }

        prefix.insert(sum);
    }

    cout << "NO";
    return 0;
}
